# Data Cleaning 
### By Swetha Revanur and Keanu Spies

Here, we implement a variety of preliminary data cleaning techniques. This notebook was adapted from [this Kaggle notebook](https://www.kaggle.com/quannguyen135/preliminary-data-cleaning-with-python). 

In [34]:
import pandas as pd
import numpy as np
import os
from datetime import timedelta, datetime

import warnings
warnings.filterwarnings('ignore')

In [35]:
base_data_dir = '../../data'
df = pd.read_csv(os.path.join(base_data_dir, 'online_sex_work.csv'), index_col = 0)
df = df.iloc[: 28831, :]

df.head()

Gender   Age Location  Verification Sexual_orientation  \
User_ID                                                          
10386.0   male  34,6        A  Non_Verified         Homosexual   
14.0      male  32,2        J  Non_Verified       Heterosexual   
16721.0   male  33,6        K  Non_Verified       Heterosexual   
16957.0   male    34        H  Non_Verified       Heterosexual   
17125.0   male  39,5        B  Non_Verified       Heterosexual   

        Sexual_polarity Looking_for Points_Rank      Last_login Member_since  \
User_ID                                                                        
10386.0          Switch         Men          50  before_10_days    17.9.2012   
14.0           Dominant       Women         518   before_1_days    1.11.2009   
16721.0        Dominant       Women         150   before_3_days     1.4.2013   
16957.0        Dominant       Women         114   before_4_days     8.4.2013   
17125.0        Dominant       Women         497   before_5_days    14.4.2013   

        Number_of_Comments_in_public_forum Time_spent_chating_H:M  \
User_ID                                                             
10386.0                                 32                    0:2   
14.0                                   710                   3:45   
16721.0                                 25                   2:15   
16957.0                                107                 359:22   
17125.0                                600                   0:21   

         Number_of_advertisments_posted  Number_of_offline_meetings_attended  \
User_ID                                                                        
10386.0                             0.0                                  0.0   
14.0                                9.0                                  0.0   
16721.0                             1.0                                  1.0   
16957.0                             1.0                                  0.0   
17125.0                             0.0                                  6.0   

         Profile_pictures                                    Friends_ID_list  \
User_ID                                                                        
10386.0               0.0                                              18260   
14.0                  0.0          11778,32024,43768,23969,27318,45884,31277   
16721.0              45.0                                  19805,21721,19802   
16957.0               1.0  40847,38183,9507,42259,5807,28118,24848,37170,...   
17125.0               8.0  1320,35739,34231,19097,20197,18069,12330,43342...   

            Risk  
User_ID           
10386.0  No_risk  
14.0     No_risk  
16721.0  No_risk  
16957.0  No_risk  
17125.0  No_risk

### Change Data Type for Some Features

In [36]:
df = df.where(df.notnull(), None)

df.index = df.index.astype(int)
df['Number_of_Comments_in_public_forum'] = df['Number_of_Comments_in_public_forum'].str.replace(' ', '').astype(int)
df['Number_of_advertisments_posted'] = df['Number_of_advertisments_posted'].astype(int)
df['Number_of_offline_meetings_attended'] = df['Number_of_offline_meetings_attended'].astype(int)
df['Profile_pictures'] = df['Profile_pictures'].astype(int)
df['Friends_ID_list'] = df['Friends_ID_list'].astype(str)
df['Risk'] = df['Risk'].astype(str)

df.head()

Gender   Age Location  Verification Sexual_orientation  \
User_ID                                                          
10386     male  34,6        A  Non_Verified         Homosexual   
14        male  32,2        J  Non_Verified       Heterosexual   
16721     male  33,6        K  Non_Verified       Heterosexual   
16957     male    34        H  Non_Verified       Heterosexual   
17125     male  39,5        B  Non_Verified       Heterosexual   

        Sexual_polarity Looking_for Points_Rank      Last_login Member_since  \
User_ID                                                                        
10386            Switch         Men          50  before_10_days    17.9.2012   
14             Dominant       Women         518   before_1_days    1.11.2009   
16721          Dominant       Women         150   before_3_days     1.4.2013   
16957          Dominant       Women         114   before_4_days     8.4.2013   
17125          Dominant       Women         497   before_5_days    14.4.2013   

         Number_of_Comments_in_public_forum Time_spent_chating_H:M  \
User_ID                                                              
10386                                    32                    0:2   
14                                      710                   3:45   
16721                                    25                   2:15   
16957                                   107                 359:22   
17125                                   600                   0:21   

         Number_of_advertisments_posted  Number_of_offline_meetings_attended  \
User_ID                                                                        
10386                                 0                                    0   
14                                    9                                    0   
16721                                 1                                    1   
16957                                 1                                    0   
17125                                 0                                    6   

         Profile_pictures                                    Friends_ID_list  \
User_ID                                                                        
10386                   0                                              18260   
14                      0          11778,32024,43768,23969,27318,45884,31277   
16721                  45                                  19805,21721,19802   
16957                   1  40847,38183,9507,42259,5807,28118,24848,37170,...   
17125                   8  1320,35739,34231,19097,20197,18069,12330,43342...   

            Risk  
User_ID           
10386    No_risk  
14       No_risk  
16721    No_risk  
16957    No_risk  
17125    No_risk

### Convert `Gender` to Binary Data

There are two genders present in this dataset (i.e., either explicitly male or female, or indirectly). Here, we will impute gender based on the `Sexual_orientation` and `Looking_for` fields, and then binarize the feature. `male` is `False` and `female` is `True.`

In [37]:
def fill_gender_na(row):
    if row['Sexual_orientation'] == 'Homosexual':
        if row['Looking_for'] == 'Men':
            return 'male'
        elif row['Looking_for'] == 'Women':
            return 'female'
    elif row['Sexual_orientation'] == 'Heterosexual':
        if row['Looking_for'] == 'Men':
            return 'female'
        elif row['Looking_for'] == 'Women':
            return 'male'
    else:
        return np.nan

In [38]:
fill_values = df.apply(fill_gender_na, axis = 1)
df['Gender'].fillna(fill_values, inplace = True)
df['Gender'].fillna(df['Gender'].mode()[0], inplace = True)

df.insert(0, 'Female', df['Gender'] == 'female')
del df['Gender']

df.head()

Female   Age Location  Verification Sexual_orientation  \
User_ID                                                          
10386    False  34,6        A  Non_Verified         Homosexual   
14       False  32,2        J  Non_Verified       Heterosexual   
16721    False  33,6        K  Non_Verified       Heterosexual   
16957    False    34        H  Non_Verified       Heterosexual   
17125    False  39,5        B  Non_Verified       Heterosexual   

        Sexual_polarity Looking_for Points_Rank      Last_login Member_since  \
User_ID                                                                        
10386            Switch         Men          50  before_10_days    17.9.2012   
14             Dominant       Women         518   before_1_days    1.11.2009   
16721          Dominant       Women         150   before_3_days     1.4.2013   
16957          Dominant       Women         114   before_4_days     8.4.2013   
17125          Dominant       Women         497   before_5_days    14.4.2013   

         Number_of_Comments_in_public_forum Time_spent_chating_H:M  \
User_ID                                                              
10386                                    32                    0:2   
14                                      710                   3:45   
16721                                    25                   2:15   
16957                                   107                 359:22   
17125                                   600                   0:21   

         Number_of_advertisments_posted  Number_of_offline_meetings_attended  \
User_ID                                                                        
10386                                 0                                    0   
14                                    9                                    0   
16721                                 1                                    1   
16957                                 1                                    0   
17125                                 0                                    6   

         Profile_pictures                                    Friends_ID_list  \
User_ID                                                                        
10386                   0                                              18260   
14                      0          11778,32024,43768,23969,27318,45884,31277   
16721                  45                                  19805,21721,19802   
16957                   1  40847,38183,9507,42259,5807,28118,24848,37170,...   
17125                   8  1320,35739,34231,19097,20197,18069,12330,43342...   

            Risk  
User_ID           
10386    No_risk  
14       No_risk  
16721    No_risk  
16957    No_risk  
17125    No_risk

### Convert European Decimal Notation in `Age`

Here, we will replace all commas (European decimal separator) with periods, while handling some unformatted values.

In [40]:
df['Age'] = df['Age'].replace('???', np.nan)
df['Age'] = df['Age'].apply(lambda x: str(x).replace(',', '.')) # str(x).split(',')[0] 
df['Age'] = df['Age'].astype(float)
df['Age'].fillna(df['Age'].mean(), inplace = True)
df['Age'].head()

User_ID
10386    34.6
14       32.2
16721    33.6
16957    34.0
17125    39.5
Name: Age, dtype: float64

### Process `Location`

Here, we replace missing values and convert remaining location names to letters. We then generate corresponding one-hot encodings.

In [41]:
df['Location'].fillna(df['Location'].mode()[0], inplace = True)
df = df.replace({'A kraj':'P', 'Jihomoravsk_ý kraj':'Q', 'South Moravian':'Q'})
df = pd.concat([df.iloc[:, :2], pd.get_dummies(df['Location']), df.iloc[:, 3:]], axis = 1)
df.head()

Female   Age    A    B    C    D    E    F    G    H   ...     \
User_ID                                                        ...      
10386    False  34.6  1.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0   ...      
14       False  32.2  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0   ...      
16721    False  33.6  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0   ...      
16957    False  34.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  1.0   ...      
17125    False  39.5  0.0  1.0  0.0  0.0  0.0  0.0  0.0  0.0   ...      

         Points_Rank      Last_login  Member_since  \
User_ID                                              
10386             50  before_10_days     17.9.2012   
14               518   before_1_days     1.11.2009   
16721            150   before_3_days      1.4.2013   
16957            114   before_4_days      8.4.2013   
17125            497   before_5_days     14.4.2013   

         Number_of_Comments_in_public_forum  Time_spent_chating_H:M  \
User_ID                                                               
10386                                    32                     0:2   
14                                      710                    3:45   
16721                                    25                    2:15   
16957                                   107                  359:22   
17125                                   600                    0:21   

         Number_of_advertisments_posted  Number_of_offline_meetings_attended  \
User_ID                                                                        
10386                                 0                                    0   
14                                    9                                    0   
16721                                 1                                    1   
16957                                 1                                    0   
17125                                 0                                    6   

         Profile_pictures                                    Friends_ID_list  \
User_ID                                                                        
10386                   0                                              18260   
14                      0          11778,32024,43768,23969,27318,45884,31277   
16721                  45                                  19805,21721,19802   
16957                   1  40847,38183,9507,42259,5807,28118,24848,37170,...   
17125                   8  1320,35739,34231,19097,20197,18069,12330,43342...   

            Risk  
User_ID           
10386    No_risk  
14       No_risk  
16721    No_risk  
16957    No_risk  
17125    No_risk  

[5 rows x 33 columns]

### Convert `Verification` to Binary Data

`Non_Verified` is `False` and `Verified` is `True.`

In [42]:
df['Verification'] = df['Verification'] != 'Non_Verified'
df['Verification'].head()

User_ID
10386    False
14       False
16721    False
16957    False
17125    False
Name: Verification, dtype: bool

### One-Hot Encoding for Categorical Fields (`Sexual_orientation`, `Sexual_polarity`, and `Looking_for`)

In [43]:
df = pd.concat([df.iloc[:, :22], pd.get_dummies(df['Looking_for']), df.iloc[:, 23:]], axis = 1)
df = pd.concat([df.iloc[:, :21], pd.get_dummies(df['Sexual_polarity']), df.iloc[:, 22:]], axis = 1)
df = pd.concat([df.iloc[:, :20], pd.get_dummies(df['Sexual_orientation']), df.iloc[:, 21:]], axis = 1)

df.columns

Index(['Female', 'Age', 'A', 'B', 'C', 'D', 'E', 'F', 'G', 'H', 'J', 'K', 'L',
       'M', 'N', 'O', 'P', 'Q', 'S', 'Verification', 'Heterosexual',
       'Homosexual', 'bicurious', 'bisexual', 'Dominant', 'Submisive',
       'Switch', 'Men', 'Men_and_Women', 'Nobody', 'Nobody_but_maybe', 'Women',
       'Points_Rank', 'Last_login', 'Member_since',
       'Number_of_Comments_in_public_forum', 'Time_spent_chating_H:M',
       'Number_of_advertisments_posted', 'Number_of_offline_meetings_attended',
       'Profile_pictures', 'Friends_ID_list', 'Risk'],
      dtype='object')

### Process `Points_Rank`

In [44]:
df['Points_Rank'].value_counts().head(10)

0     19931
15     3995
5       427
30      333
20      275
1       248
a       209
10      195
16      169
50      167
Name: Points_Rank, dtype: int64

Aside from integers, the `Points_Rank` column contains 209 entries with `a` as their data. While it is not clear what `a` represents, 209 is relatively small compared to our dataset, so we replace all instances of `a` with the mode (0), and convert the column to integer datatype.

In [45]:
df['Points_Rank'] = df['Points_Rank'].str.replace(' ', '')
df['Points_Rank'].replace(to_replace = 'a', value = '0', inplace = True)
df['Points_Rank'] = df['Points_Rank'].astype(int)

df['Points_Rank'].head()

User_ID
10386     50
14       518
16721    150
16957    114
17125    497
Name: Points_Rank, dtype: int64

### Convert `Last_login` strings to number of days (integers)

In [46]:
df['Last_login'] = df['Last_login'].apply(lambda x: x.split('_')[1]).astype(int)
df['Last_login'].head()

User_ID
10386    10
14        1
16721     3
16957     4
17125     5
Name: Last_login, dtype: int64

### Process `Member_since`

We convert data in the `Member_since` column to Python's `datetime` datatype. An entry was evidently inputted incorrectly (it contains 0,278159722), so we replace it with the mode.

Some entries contain `dnes`, which translates "today" in Czech (probably where the dataset originates from), so we arbitrarily replace them with the maximum date among the other entries, with 1 more day added.

Finally, we also convert the date to the number of days since joining.

In [47]:
df['Member_since'].replace(to_replace = '0,278159722', value = df['Member_since'].mode()[0], inplace = True)
df['Member_since'].replace(to_replace = 'dnes', value = np.nan, inplace = True)
df['Member_since'] = pd.to_datetime(df['Member_since'], format = '%d.%m.%Y')
df['Member_since'] = df['Member_since'].fillna(df['Member_since'].max() + timedelta(days = 1))

def get_day_diff(row):
    return (datetime.now() - row).days
df['Member_since'] = df['Member_since'].apply(get_day_diff)

df['Member_since'].head()

User_ID
10386    2252
14       3303
16721    2056
16957    2049
17125    2043
Name: Member_since, dtype: int64

### Process `Time_spent_chating_H:M`

We will replace data in this column with the total number of minutes indicated. Some entries have the `H:M` (hour:minute) format, while others have what we suspect to be the `D:H:M` (day:hour:minute) format, so we define the `get_n_minutes` function to handle that.

In [48]:
def get_n_minutes(row):
    time_components = row.split(':')
    if len(time_components) == 2:
        return int(time_components[0]) * 60 + int(time_components[1])
    elif len(time_components) == 3:
        return int(time_components[0]) * 1440 + int(time_components[1]) * 60 + int(time_components[2])

df['Time_spent_chating_H:M'] = df['Time_spent_chating_H:M'].str.replace(' ', '')
df['Time_spent_chating_H:M'] = df['Time_spent_chating_H:M'].apply(get_n_minutes)

df['Time_spent_chating_H:M'].head()

User_ID
10386        2
14         225
16721      135
16957    21562
17125       21
Name: Time_spent_chating_H:M, dtype: int64

### Add `Number_of_friends` feature

We add a column to indicate the length of the `Friends_ID_List` field for each user.

In [49]:
def get_n_friends(row):
    friend_ids = row.split(',')
    if friend_ids[0] == 'None':
        return 0
    return len(friend_ids)

df.insert(40, 'Number_of_friends', df['Friends_ID_list'].apply(get_n_friends))
df.iloc[:5, 20:]

Heterosexual  Homosexual  bicurious  bisexual  Dominant  Submisive  \
User_ID                                                                       
10386             0.0         1.0        0.0       0.0       0.0        0.0   
14                1.0         0.0        0.0       0.0       1.0        0.0   
16721             1.0         0.0        0.0       0.0       1.0        0.0   
16957             1.0         0.0        0.0       0.0       1.0        0.0   
17125             1.0         0.0        0.0       0.0       1.0        0.0   

         Switch  Men  Men_and_Women  Nobody   ...     Last_login  \
User_ID                                       ...                  
10386       1.0  1.0            0.0     0.0   ...             10   
14          0.0  0.0            0.0     0.0   ...              1   
16721       0.0  0.0            0.0     0.0   ...              3   
16957       0.0  0.0            0.0     0.0   ...              4   
17125       0.0  0.0            0.0     0.0   ...              5   

         Member_since  Number_of_Comments_in_public_forum  \
User_ID                                                     
10386            2252                                  32   
14               3303                                 710   
16721            2056                                  25   
16957            2049                                 107   
17125            2043                                 600   

         Time_spent_chating_H:M  Number_of_advertisments_posted  \
User_ID                                                           
10386                         2                               0   
14                          225                               9   
16721                       135                               1   
16957                     21562                               1   
17125                        21                               0   

         Number_of_offline_meetings_attended  Profile_pictures  \
User_ID                                                          
10386                                      0                 0   
14                                         0                 0   
16721                                      1                45   
16957                                      0                 1   
17125                                      6                 8   

         Number_of_friends                                    Friends_ID_list  \
User_ID                                                                         
10386                    1                                              18260   
14                       7          11778,32024,43768,23969,27318,45884,31277   
16721                    3                                  19805,21721,19802   
16957                   12  40847,38183,9507,42259,5807,28118,24848,37170,...   
17125                   35  1320,35739,34231,19097,20197,18069,12330,43342...   

            Risk  
User_ID           
10386    No_risk  
14       No_risk  
16721    No_risk  
16957    No_risk  
17125    No_risk  

[5 rows x 23 columns]

### Process `Risk`

For each entry, we replace it with 0 if it says `No_risk`, 1 if `High_risk`, and `np.nan` if `unknown_risk`.

In [50]:
def get_risk(row):
    if row == 'No_risk':
        return 0
    elif row == 'High_risk':
        return 1
    
    return np.nan

df['Risk'] = df['Risk'].apply(get_risk)
df['Risk'].head()

User_ID
10386    0.0
14       0.0
16721    0.0
16957    0.0
17125    0.0
Name: Risk, dtype: float64

### Process Column Names

Here, we fix typos and perform casefolding on the column names.

In [51]:
df.columns = df.columns.str.lower()
df = df.rename(columns = {'submisive': 'submissive', 
                          'time_spent_chating_h:m': 'number_of_minutes_spent_chatting', 
                          'number_of_advertisments_posted': 'number_of_advertisements_posted', 
                          'profile_pictures': 'number_of_profile_pictures'})

### Save and Export

In [52]:
df = df.sort_values(by = ['risk'], na_position = 'last')
df = df.replace({'None': None})
df = df * 1 # replaces T/F with 1/0
df = df.where(df.notnull(), None)

df.to_pickle(os.path.join(base_data_dir, 'online_sex_work_clean.pkl'))

In [53]:
df

female   age  a  b  c  d  e  f  g  h  ...  last_login member_since  \
User_ID                                        ...                            
10386         0  34.6  1  0  0  0  0  0  0  0  ...          10         2252   
37521         0    62  0  0  0  0  0  0  0  0  ...           1         1157   
3987          0  28.1  0  0  0  0  0  0  0  0  ...           9         2557   
40412         0  24.6  0  0  0  0  1  0  0  0  ...           4          994   
42646         0  22.1  0  0  0  0  1  0  0  0  ...           1          868   
43605         1  24.9  0  0  0  0  0  0  0  1  ...           3          816   
43656         0  54.3  0  0  0  0  0  0  0  0  ...           3          813   
44151         0  29.9  0  0  0  0  0  0  1  0  ...           5          789   
45230         0  24.5  0  0  0  0  1  0  0  0  ...           1          734   
45884         1  26.4  0  0  0  0  0  0  0  0  ...           1          697   
46195         1  28.6  0  1  0  0  0  0  0  0  ...           1          681   
46629         0  24.6  0  0  0  0  1  0  0  0  ...           1          661   
46681         0  42.6  0  0  0  0  1  0  0  0  ...          30          658   
46911         1  20.7  0  1  0  0  0  0  0  0  ...           3          646   
47917         0  35.5  0  1  0  0  0  0  0  0  ...           3          598   
48294         0    24  1  0  0  0  0  0  0  0  ...           1          581   
6035          0  52.3  1  0  0  0  0  0  0  0  ...           1         2329   
6065          0  47.2  0  0  0  0  0  1  0  0  ...           1         2329   
6163          0  31.4  0  0  1  0  0  0  0  0  ...           1         2328   
6418          0    41  0  0  0  0  0  1  0  0  ...           1         2326   
6732          0    36  0  0  0  0  0  0  0  0  ...           7         2322   
83            0  28.7  0  1  0  0  0  0  0  0  ...           3         3295   
8696          0  41.7  1  0  0  0  0  0  0  0  ...           3         2292   
881           0  40.8  0  0  0  0  0  0  0  0  ...           1         3121   
9479          0  31.8  1  0  0  0  0  0  0  0  ...          64         2275   
9550          1  49.9  0  1  0  0  0  0  0  0  ...          89         2273   
37385         0  35.6  0  0  1  0  0  0  0  0  ...           1         1165   
35266         0  39.2  0  0  0  0  0  0  0  0  ...          11         1277   
47601         0  40.4  0  0  0  0  0  0  0  0  ...          16          612   
34759         0  26.8  0  0  1  0  0  0  0  0  ...           1         1307   
...         ...   ... .. .. .. .. .. .. .. ..  ...         ...          ...   
9923          0  46.3  0  0  0  0  0  0  0  1  ...        1656         2265   
9925          1  32.1  1  0  0  0  0  0  0  0  ...          97         2265   
9926          0  45.6  1  0  0  0  0  0  0  0  ...        1697         2265   
9928          0  36.1  0  0  0  0  1  0  0  0  ...        1687         2264   
9929          0  26.8  1  0  0  0  0  0  0  0  ...        1697         2264   
9931          0  56.3  0  0  0  0  0  0  0  0  ...        1697         2264   
9932          0  43.7  1  0  0  0  0  0  0  0  ...         979         2264   
9934          1  40.9  0  0  0  0  0  0  1  0  ...        1691         2264   
9937          0  47.2  1  0  0  0  0  0  0  0  ...         884         2264   
9938          1  44.7  1  0  0  0  0  0  0  0  ...        1349         2264   
9939          0  40.1  1  0  0  0  0  0  0  0  ...        1697         2264   
994           0  32.7  0  0  0  0  0  0  0  0  ...         205         3104   
9940          0  28.4  1  0  0  0  0  0  0  0  ...         944         2264   
9941          0    33  1  0  0  0  0  0  0  0  ...          12         2264   
9942          0  26.8  1  0  0  0  0  0  0  0  ...         453         2264   
9943          0  30.2  1  0  0  0  0  0  0  0  ...        1697         2264   
9944          0    23  0  0  0  0  1  0  0  0  ...        1510         2264   
9945          0  48.7  0  1  0  0  0  0  0  0  ...         